File: Chat-and-Video-Interview-with-Sentiment-Analysis-and-Avatar-Conversational-Al--Beta-main/talbotiq-platform/server/routes/faceCache.ts
import { Router } from 'express'
import { createHash } from 'node:crypto'
import fs from 'node:fs'
import path from 'node:path'
import { fileURLToPath } from 'node:url'
import { ah, HttpError } from '../util/ah'
import { bearerToken, contextFromToken } from '../middleware/auth'

/**
 * Replica-preview cache — makes the face picker instant.
 *
 * Tavus replica previews are MP4s on a remote CDN, so every tile used to buffer
 * over the network the moment it scrolled into view. This route downloads each
 * preview ONCE, persists it under server/data/face-cache/ (survives restarts;
 * the dir is gitignored), and serves it from local disk with immutable cache
 * headers thereafter. The client warms the cache in the background right after
 * the replica list loads, so the first picker open is already local-fast.
 *
 * Not an open proxy: https-only, and the upstream host must match a small
 * allowlist (Tavus + its CDNs; extend via FACE_CACHE_HOSTS). Auth: recruiter
 * only — accepts the bearer header or ?token= because <video> tags cannot send
 * headers (same pattern as the WebSocket upgrades).
 */

const here = path.dirname(fileURLToPath(import.meta.url))
const CACHE_DIR = path.join(here, '..', 'data', 'face-cache')

const DEFAULT_HOSTS = ['tavus.io', 'tavusapi.com', 'tavus.video', 'cloudfront.net', 'amazonaws.com']
function allowedHost(hostname: string): boolean {
  const extra = (process.env.FACE_CACHE_HOSTS ?? '')
    .split(/[\s,]+/).map((s) => s.trim().toLowerCase()).filter(Boolean)
  const h = hostname.toLowerCase()
  return [...DEFAULT_HOSTS, ...extra].some((d) => h === d || h.endsWith(`.${d}`))
}

// Some Tavus stock previews exceed 40 MB (observed on cdn.replica.tavus.io), so
// the cap is generous — instant playback is worth the disk. Still bounded so a
// mistaken URL can't fill the drive.
const MAX_BYTES = 150 * 1024 * 1024

/** In-flight downloads keyed by cache key, so N tiles asking for the same video
 *  while it's still downloading share one upstream fetch. */
const inflight = new Map<string, Promise<string>>()

async function ensureCached(url: string): Promise<string> {
  const key = createHash('sha1').update(url).digest('hex')
  const file = path.join(CACHE_DIR, `${key}.mp4`)
  if (fs.existsSync(file)) return file
  const running = inflight.get(key)
  if (running) return running

  const download = (async () => {
    fs.mkdirSync(CACHE_DIR, { recursive: true })
    const r = await fetch(url, { redirect: 'follow', signal: AbortSignal.timeout(30_000) })
    if (!r.ok) throw new HttpError(502, `Upstream fetch failed (${r.status})`)
    const declared = Number(r.headers.get('content-length') ?? 0)
    if (declared > MAX_BYTES) throw new HttpError(502, 'Preview too large to cache')
    const buf = Buffer.from(await r.arrayBuffer())
    if (buf.byteLength > MAX_BYTES) throw new HttpError(502, 'Preview too large to cache')
    // Atomic write: never leave a half-downloaded file where sendFile can find it.
    const tmp = `${file}.${process.pid}.${Date.now()}.tmp`
    fs.writeFileSync(tmp, buf)
    fs.renameSync(tmp, file)
    return file
  })().finally(() => inflight.delete(key))

  inflight.set(key, download)
  return download
}

export const faceCacheRouter = Router()

// Recruiter auth via header OR ?token= (media elements can't set headers).
faceCacheRouter.use((req, _res, next) => {
  Promise.resolve()
    .then(async () => {
      const token = bearerToken(req) ?? (typeof req.query.token === 'string' ? req.query.token : null)
      if (!token) throw new HttpError(401, 'Authentication required')
      const ctx = await contextFromToken(token)
      if (ctx.role !== 'recruiter') throw new HttpError(403, 'Recruiter access required')
      req.auth = ctx
      next()
    })
    .catch(next)
})

// GET /api/avatar/face-cache?url=<https CDN url> → the cached MP4 (Range-capable).
// HEAD warms the cache without transferring the body (used by the client warmer).
faceCacheRouter.get('/', ah(async (req, res) => {
  const url = typeof req.query.url === 'string' ? req.query.url : ''
  let parsed: URL
  try { parsed = new URL(url) } catch { throw new HttpError(400, 'Invalid url') }
  if (parsed.protocol !== 'https:' || !allowedHost(parsed.hostname))
    throw new HttpError(400, 'Host not allowed')

  try {
    const file = await ensureCached(url)
    res.sendFile(file, { maxAge: '365d', immutable: true })
  } catch (err) {
    // Graceful degradation: if the download fails, send the browser straight to
    // the CDN so the picker still shows the face (just not from cache).
    console.warn('[face-cache] falling back to CDN for', parsed.hostname, '-', err instanceof Error ? err.message : err)
    res.redirect(302, url)
  }
}))

File: backend/app/web/routes/face_cache.py
"""Replica-preview cache — ports `server/routes/faceCache.ts`.

Tavus replica previews are MP4s on a remote CDN, so every tile in the face picker used
to buffer over the network the moment it scrolled into view. This fetches each preview
once, keeps it in Firebase Storage, and redirects the browser to a URL with immutable
cache headers — so the second look is served from the browser's own cache and the
picker opens instantly.

**Stored in the bucket, never on the server's filesystem.** The Express version wrote
MP4s under `server/data/face-cache/`, which cannot work here: a container's disk is
ephemeral, so the cache would be lost on every deploy, and each worker would keep its
own copy while missing what the others had already fetched.

**Not an open proxy.** HTTPS only, and the upstream host must match a small allowlist —
without it, an authenticated caller could use this endpoint to fetch arbitrary URLs from
inside the network and have the bytes handed back.

A cache miss is not a failure: if the fetch or the upload fails, the browser is
redirected to the original CDN URL and the picker still shows the face.
"""

from __future__ import annotations

import asyncio
import hashlib
import logging
from urllib.parse import urlparse

from fastapi import APIRouter, HTTPException, Query, Request, status
from fastapi.responses import RedirectResponse

from app.config import Settings
from app.providers.base import http_client
from app.security import AuthedUser
from app.web.deps import WebUserFromQuery, settings_of
from app.web.services import storage

logger = logging.getLogger("web.face_cache")

# Mounted ahead of /avatar in app.web so this more specific path wins, mirroring the
# Express mount order.
router = APIRouter(prefix="/avatar/face-cache", tags=["web:avatar"])

DEFAULT_HOSTS = (
    "tavus.io",
    "tavusapi.com",
    "tavus.video",
    "cloudfront.net",
    "amazonaws.com",
)

# Some Tavus stock previews exceed 40 MB (observed on cdn.replica.tavus.io), so the cap
# is generous — instant playback is worth the storage. Still bounded, so a mistaken URL
# cannot fill the bucket.
MAX_BYTES = 150 * 1024 * 1024

DOWNLOAD_TIMEOUT_SECONDS = 30.0

# Where cached previews live. Prefixed like every other web-owned collection, so it is
# obvious in the console which surface owns them.
OBJECT_PREFIX = "web_face_cache"

# Fetches in progress, keyed by object path, so N tiles asking for the same video share
# one upstream fetch instead of racing to upload it.
_inflight: dict[str, asyncio.Task] = {}


def allowed_hosts(settings: Settings) -> tuple[str, ...]:
    extra = tuple(
        part.strip().lower()
        for part in (settings.face_cache_hosts or "").replace(",", " ").split()
        if part.strip()
    )
    return DEFAULT_HOSTS + extra


def is_allowed(settings: Settings, hostname: str) -> bool:
    """Exact match or a subdomain of an allowed host.

    The leading dot in the suffix check is what stops `nottavus.io` matching `tavus.io` —
    a bare `endswith("tavus.io")` would accept an attacker's domain.
    """
    host = (hostname or "").lower()
    if not host:
        return False
    return any(
        host == domain or host.endswith(f".{domain}") for domain in allowed_hosts(settings)
    )


def cache_key(url: str) -> str:
    """A stable key for a URL.

    A hash, so no part of the URL reaches the object path: a CDN path could contain `..`
    or a separator and land the object somewhere unintended.
    """
    return hashlib.sha256(url.encode()).hexdigest()


def object_path(url: str) -> str:
    return f"{OBJECT_PREFIX}/{cache_key(url)}.mp4"


def download_token(url: str) -> str:
    """A deterministic token for the cached object's URL.

    Derived from the URL rather than randomly generated so the download URL is
    reproducible without a second round trip to read the object's metadata. Safe here
    because the content is a PUBLIC replica preview from a CDN — the token gates nothing
    that was not already public, and the host allowlist is what stops this endpoint being
    used to cache anything else.
    """
    return hashlib.sha256(f"face-cache:{url}".encode()).hexdigest()[:32]


async def _fetch_and_store(settings: Settings, url: str) -> str:
    """Fetch a preview from the CDN and put it in the bucket. Returns its URL."""
    response = await http_client().get(
        url, follow_redirects=True, timeout=DOWNLOAD_TIMEOUT_SECONDS
    )
    if response.status_code >= 400:
        raise RuntimeError(f"Upstream fetch failed ({response.status_code})")

    declared = int(response.headers.get("content-length") or 0)
    if declared > MAX_BYTES or len(response.content) > MAX_BYTES:
        raise RuntimeError("Preview too large to cache")

    return await storage.upload(
        settings,
        object_path(url),
        response.content,
        content_type=response.headers.get("content-type") or "video/mp4",
        token=download_token(url),
    )


async def ensure_cached(settings: Settings, url: str) -> str:
    """The bucket URL for a preview, fetching it once if it is not there yet."""
    existing = await storage.find(settings, object_path(url))
    if existing:
        return existing

    key = object_path(url)
    running = _inflight.get(key)
    if running is not None:
        return await running

    task = asyncio.ensure_future(_fetch_and_store(settings, url))
    _inflight[key] = task
    try:
        return await task
    finally:
        _inflight.pop(key, None)


@router.api_route("", methods=["GET", "HEAD"], summary="Serve a cached replica preview")
async def serve(
    request: Request,
    url: str = Query(default=""),
    user: AuthedUser = WebUserFromQuery,
):
    """Redirect to the cached preview, fetching it into the bucket on first ask.

    Authenticated via the bearer header OR `?token=`, because a `<video>` element cannot
    set headers. HEAD warms the cache without transferring the body, which is what the
    client's background warmer uses.

    ROLE-GATING: the Express route required the recruiter role. Role gating is deferred,
    so any authenticated caller is accepted — the host allowlist is what keeps this from
    being an open proxy.
    """
    settings = settings_of(request)

    parsed = urlparse(url)
    if not parsed.scheme or not parsed.netloc:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, "Invalid url")
    if parsed.scheme != "https" or not is_allowed(settings, parsed.hostname or ""):
        raise HTTPException(status.HTTP_400_BAD_REQUEST, "Host not allowed")

    try:
        cached = await ensure_cached(settings, url)
    except Exception as exc:  # noqa: BLE001 - a cache miss is not a failure
        # Send the browser to the CDN instead, so the picker still shows the face — just
        # not from cache. The URL is already allowlisted, so this cannot redirect anywhere
        # new.
        logger.warning("face-cache falling back to CDN for %s — %s", parsed.hostname, exc)
        return RedirectResponse(url, status_code=status.HTTP_302_FOUND)

    # 302, not a proxied stream: the bytes go browser↔Google directly, so this service
    # never carries video, and the immutable cache headers on the object mean the second
    # look never comes back here at all.
    return RedirectResponse(cached, status_code=status.HTTP_302_FOUND)
